### ISSUE TITLE
Build virtual Zarr store using Xarray's dataset.to_zarr(region="...") model

### ISSUE THREAD

**Issue by maxrjones** [zarr-developers/VirtualiZarr#308]
I'm wondering how we can build virtualization pipelines as a map rather than a map-reduce process. The map paradigm would follow the structure below from [Earthmover's excellent blog post on serverless datacube pipelines](https://earthmover.io/blog/serverless-datacube-pipeline), where the virtual dataset from each worker would get written directly to an Icechunk Virtual Store rather than being transferred back to the coordination node for a concatenation step. Similar to their post, the parallelization between workers during virtualization could leverage a serverless approach like lithops or coiled. Dask concurrency could speed up virtualization within each worker. I think the main feature needed for this to work is an analogue to `xr.Dataset.to_zarr(region="...")`, complementary to `xr.Dataset.to_zarr(append_dim="...")`(documented in https://docs.xarray.dev/en/latest/generated/xarray.Dataset.to_zarr.html)  (xref https://github.com/zarr-developers/VirtualiZarr/issues/21, https://github.com/zarr-developers/VirtualiZarr/pull/272).

![Image](https://github.com/user-attachments/assets/4886d738-dfba-4c3a-83f3-24640aef2607)

I tried to check that this feature request doesn't already exist, but apologies if I missed something.

**Comment by TomNicholas** [comment]
This is a really extremely good question that I have also thought about a bit.

I think a big difference is that this "map" approach assumes more knowledge the structure of the dataset up front. To use zarr regions you have to know how big the regions are in advance to create an empty zarr array that will be filles, but In the "map-reduce" approach you are able to concatenate files whose shapes you don't know until you open them. Whether that actually matters much in practice is another question.

One thing to note is that IIUC having each worker write a region of virtual references would mean one icechunk commit per region, whereas transferring everything back to one node means one commit per logical version of the dataset, which seems a little nicer.

>  I think the main feature needed for this to work is an analogue to `xr.Dataset.to_zarr(region="...")`

I agree.

**Comment by norlandrhagen** [comment]
I really like the idea behind this approach @maxrjones! Being able to skip transferring all the references back to a 'reduce' step seem great. Also wondering about how all the commits between workers would work with icechunk.

**Comment by norlandrhagen** [comment]
This also seems like a good way to protect against a single failed reference killing the entire job.

**Comment by dcherian** [comment]
> I think a big difference is that this "map" approach assumes more knowledge the structure of the dataset up front. 

:+1:

> IIUC having each worker write a region of virtual references would mean one icechunk commit per region, 

It does not, once https://github.com/earth-mover/icechunk/pull/357 is in

**Comment by maxrjones** [comment]
> > IIUC having each worker write a region of virtual references would mean one icechunk commit per region,

> It does not, once https://github.com/earth-mover/icechunk/pull/357 is in. Though in practice, that also does map-reduce on changesets which will contain virtual references in this case...

Thanks for this information! Do you think it's accurate that map-reduce on changesets would still be more performant than map-reduce on virtual datasets when operating on thousands or more of files, such that this feature would be helpful despite not truly by a map operation? Also, do you anticipate there being other prerequisites in addition to your PR before VirtualiZarr could have something like `dataset_to_icechunk(region="...")`?

**Comment by dcherian** [comment]
I've been thinking that a "CropCodec" might work for uniform overlaps.

Here is an example of a TIFF dataset with no overlaps if anyone is interested: https://icechunk.io/en/latest/ingestion/glad-ingest/

**Comment by TomNicholas** [comment]
I agree with all of that. I also use the pattern of appending and committing batches in a loop, but I agree that is really just a special case of writing to regions then committing in a loop.

I don't think it's a bad thing that there is a limit to the number of refs you can commit in one go, it's kind of crazy to commit many many millions of refs in one go - at that scale you should be batching ingest for other reliability reasons anyway.

> I am willing to submit a PR if you are interested and guide me how you like the code be structured (e.g. whether to depend on Xarray or vendor the logic)

That would be awesome! Whether or not we vendor xarray kind of depends on how much of xarray we would have to vendor... Some of the internals are more stable than others. I would start out by not vendoring anything and we'll see how fragile that looks.
